src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.Tree.cs

```diff
@@ -0,0 +1,505 @@
+                public ArrayBuilder<Closure> Closures { get; } = ArrayBuilder<Closure>.GetInstance();
+
+                /// <summary>
+                /// A list of all locals or parameters that were declared in this scope and captured
+                /// in nested scopes. "Declared" refers to the start of the variable lifetime (which,
+                /// at this point in lowering, should be equivalent to lexical scope).
+                /// </summary>
+                public ArrayBuilder<Symbol> DeclaredVariables { get; } = ArrayBuilder<Symbol>.GetInstance();
+
+                /// <summary>
+                /// The bound node representing this scope. This roughly corresponds to the bound
+                /// node for the block declaring locals for this scope, although parameters of
+                /// methods/closures are introduced into their Body's scope and do not get their
+                /// own scope.
+                /// </summary>
+                public BoundNode BoundNode { get; }
+
+                /// <summary>
+                /// The closure that this scope is nested inside. Null if this scope is not nested
+                /// inside a closure.
+                /// </summary>
+                public Closure ContainingClosure { get; }
+
+                public Scope(Scope parent, BoundNode boundNode, Closure containingClosure)
+                {
+                    Debug.Assert(boundNode != null);
+
+                    Parent = parent;
+                    BoundNode = boundNode;
+                    ContainingClosure = containingClosure;
+                }
+
+                public void Free()
+                {
+                    foreach (var scope in NestedScopes)
+                    {
+                        scope.Free();
+                    }
+                    NestedScopes.Free();
+
+                    foreach (var closure in Closures)
+                    {
+                        closure.Free();
+                    }
+                    Closures.Free();
+                    DeclaredVariables.Free();
+                }
+
+                public override string ToString() => BoundNode.Syntax.GetText().ToString();
+            }
+
+            /// <summary>
+            /// The Closure type represents a lambda or local function and stores
+            /// information related to that closure. After initially building the
+            /// <see cref="Scope"/> tree the only information available is
+            /// <see cref="OriginalMethodSymbol"/> and <see cref="CapturedVariables"/>.
+            /// Subsequent passes are responsible for translating captured
+            /// variables into captured environments and for calculating
+            /// the rewritten signature of the method.
+            /// </summary>
+            private sealed class Closure
+            {
+                /// <summary>
+                /// The method symbol for the original lambda or local function.
```

Comment:
I think the literature usually refers to that as a "capture environment" or "closure environment".